server/api/src/rsGeneralAdmin.cpp

```diff
@@ -52,6 +53,58 @@ namespace
         return std::find_if(b, e, [](unsigned char _ch) { return ::isspace(_ch); }) != e;
     } // contains_whitespace
 
+     auto throw_if_user_type_can_result_in_non_functional_irods_server(
+        RsComm& rsComm,
+        const std::string_view _option,
+        const std::string_view _user_name,
+        const std::string_view _new_user_type
+    ) -> void {
+            if ("rodsadmin" == _new_user_type ||
+                "type"      == _option
+               ) { return; }
+
+            std::set<std::string> server_administrators_in_zone;
+
+            BytesBuf* bbuf = nullptr;
+            irods::at_scope_exit free_buf { [&bbuf] { std::free(bbuf); }  };
+
+            if (const auto ec = rsZoneReport(&rsComm, &bbuf); ec < 0 ) {
+                const auto msg = "The server experienced an unexpected error while processing the request. Please report to the administrator.";
+                addRErrorMsg(&rsComm.rError, SYS_NOT_ALLOWED, msg);
+                THROW(ec, msg);
+            }
+
+            const auto buf = static_cast<char*> (bbuf->buf);
+            const nlohmann::json zone_report = nlohmann::json::parse(buf, buf + bbuf->len);
+            const auto& zones = zone_report.at("zones");
+
+            const auto * local_zone_name = getLocalZoneName();
+
+            for ( const auto& zone : zones ) {
+                const auto icat_server_env = zone.at("icat_server").at("serivce_account_environment");
+
+                if ( icat_server_env.at("irods_zone_name").get_ref<const std::string&>() != local_zone_name ) {
+                    continue;
+                }
+
+                server_administrators_in_zone.insert(
+                    icat_server_env.at("irods_user_name").get<std::string>()
+                );
+
+                for ( const auto& server : zone.at("servers") ) {
+                    server_administrators_in_zone.insert(
+                        server.at("service_account_environment").at("irods_user_name").get<std::string>()
+                    );
+                }
+            }
+
+            if ( server_administrators_in_zone.find(_user_name.data()) != server_administrators_in_zone.end() ) {
+                const auto msg = "ERROR: Cannot downgrade another rodsadmin running another server in this zone.";
```

Comment:
consider putting the attempted username in the error message?  and the other servername... in case that helps the admin of a large zone.

Reply:
Funnily enough, I actually did that before second-guessing myself. Whoops.